diff --git a/edu.harvard.i2b2.server-common/src/core/edu/harvard/i2b2/common/util/axis2/ServiceClient.java b/edu.harvard.i2b2.server-common/src/core/edu/harvard/i2b2/common/util/axis2/ServiceClient.java
index ea5d545..3b3b78a 100755
--- a/edu.harvard.i2b2.server-common/src/core/edu/harvard/i2b2/common/util/axis2/ServiceClient.java
+++ b/edu.harvard.i2b2.server-common/src/core/edu/harvard/i2b2/common/util/axis2/ServiceClient.java
@@ -84,6 +84,7 @@ public class ServiceClient {
 			options.setProperty(Constants.Configuration.ENABLE_REST, Constants.VALUE_TRUE);
 			options.setProperty(HTTPConstants.CACHED_HTTP_CLIENT, httpClient);	
 			options.setProperty(HTTPConstants.REUSE_HTTP_CLIENT, Constants.VALUE_TRUE);
+			options.setTimeOutInMilliSeconds(43200000L);
 			serviceClient.setOptions(options);
 			
 			OMElement result = serviceClient.sendReceive(request);
@@ -248,4 +249,4 @@ public class ServiceClient {
 		return serviceClient;
 	}
 
-}
\ No newline at end of file
+}
