diff --git a/edu.harvard.i2b2.ontology/src/core/edu/harvard/i2b2/ontology/dao/ConceptPersistDao.java b/edu.harvard.i2b2.ontology/src/core/edu/harvard/i2b2/ontology/dao/ConceptPersistDao.java
index 71c4224..941ae7f 100755
--- a/edu.harvard.i2b2.ontology/src/core/edu/harvard/i2b2/ontology/dao/ConceptPersistDao.java
+++ b/edu.harvard.i2b2.ontology/src/core/edu/harvard/i2b2/ontology/dao/ConceptPersistDao.java
@@ -929,7 +929,7 @@ public class ConceptPersistDao extends JdbcDaoSupport {
 					"(c_hlevel,c_fullname,c_name,c_synonym_cd,c_visualattributes,c_basecode,c_facttablecolumn," +
 					"c_totalnum, c_metadataxml, c_tablename,c_columnname,c_columndatatype,c_operator,c_dimcode,c_comment,c_tooltip," +
 					"import_date, download_date, update_date, sourcesystem_cd, valuetype_cd, m_applied_path, m_exclusion_cd, c_path, c_symbol, pcori_basecode) "+
-					"VALUES (?, ?, ?, ?, ?, ?, ?, 	?, ?,?, ?, ?, ?,?, ?, ?,    ?, ?, ?, ?, ?, ?, ?, ?,?, ?)";
+					"VALUES (?, ?, ?, ?, ?, ?, ?, 	?, ?,?, ?, ?, ?,?, ?, ?,    ?, ?, ?, ?, ?, ?, ?, ?,?, ?) ON CONFLICT DO NOTHING";
 
 			log.info(startSql);
 			
